txnkv/transaction/pessimistic.go

```diff
@@ -150,156 +168,378 @@ func (action actionPessimisticLock) handleSingleBatch(c *twoPhaseCommitter, bo *
+		}
+		return true, nil
+	}
+	var locks []*txnlock.Lock
+	for _, keyErr := range keyErrs {
+		// Check already exists error
+		if alreadyExist := keyErr.GetAlreadyExist(); alreadyExist != nil {
+			e := &tikverr.ErrKeyExist{AlreadyExist: alreadyExist}
+			return true, c.extractKeyExistsErr(e)
+		}
+		if deadlock := keyErr.Deadlock; deadlock != nil {
+			return true, errors.WithStack(&tikverr.ErrDeadlock{Deadlock: deadlock})
+		}
 
-			// Handle the case that the TiKV's version is too old and doesn't support `CheckExistence`.
-			// If `CheckExistence` is set, `ReturnValues` is not set and `CheckExistence` is not supported, skip
-			// retrieving value totally (indicated by `skipRetrievingValue`) to avoid panicking.
-			skipRetrievingValue := !action.ReturnValues && action.CheckExistence && len(lockResp.NotFounds) == 0
+		// Extract lock from key error
+		lock, err1 := txnlock.ExtractLockFromKeyErr(keyErr)
+		if err1 != nil {
+			return true, err1
+		}
+		locks = append(locks, lock)
+	}
 
-			if (action.ReturnValues || action.CheckExistence) && !skipRetrievingValue {
+	// Because we already waited on tikv, no need to Backoff here.
+	// tikv default will wait 3s(also the maximum wait value) when lock error occurs
+	if diagCtx.resolvingRecordToken == nil {
+		token := c.store.GetLockResolver().RecordResolvingLocks(locks, c.startTS)
+		diagCtx.resolvingRecordToken = &token
+	} else {
+		c.store.GetLockResolver().UpdateResolvingLocks(locks, c.startTS, *diagCtx.resolvingRecordToken)
+	}
+	resolveLockOpts := txnlock.ResolveLocksOptions{
+		CallerStartTS: 0,
+		Locks:         locks,
+	}
+	if action.LockCtx.Stats != nil {
+		resolveLockOpts.Detail = &action.LockCtx.Stats.ResolveLock
+	}
+	resolveLockRes, err := c.store.GetLockResolver().ResolveLocksWithOpts(bo, resolveLockOpts)
+	if err != nil {
+		return true, err
+	}
+
+	// If msBeforeTxnExpired is not zero, it means there are still locks blocking us acquiring
+	// the pessimistic lock. We should return acquire fail with nowait set or timeout error if necessary.
+	if resolveLockRes.TTL > 0 {
+		if action.LockWaitTime() == kv.LockNoWait {
+			return true, errors.WithStack(tikverr.ErrLockAcquireFailAndNoWaitSet)
+		} else if action.LockWaitTime() == kv.LockAlwaysWait {
+			// do nothing but keep wait
+		} else {
+			// the lockWaitTime is set, we should return wait timeout if we are still blocked by a lock
+			if time.Since(action.WaitStartTime).Milliseconds() >= action.LockWaitTime() {
+				return true, errors.WithStack(tikverr.ErrLockWaitTimeout)
+			}
+		}
+		if action.LockCtx.PessimisticLockWaited != nil {
+			atomic.StoreInt32(action.LockCtx.PessimisticLockWaited, 1)
+		}
+	}
+
+	return false, nil
+}
+
+func (action actionPessimisticLock) handlePessimisticLockResponseForceLockMode(c *twoPhaseCommitter, bo *retry.Backoffer, batch *batchMutations, mutationsPb []*kvrpcpb.Mutation, resp *tikvrpc.Response, diagCtx *diagnosticContext) (finished bool, retryMutations CommitterMutations, err error) {
```

Comment:
If a key is locked forcibly, how does the caller know about it considering the returned results and retry the whole statement if necessary?

Reply:
The LockedWithConflictTS information will be stored in the LockCtx like how we store returned values in it, and TiDB will check it.